text/3730-nit-lint-level.md

```diff
@@ -0,0 +1,402 @@
+
+## Rustc
+
+Rustc focuses on the mechanics of a new, visible lint level `nit`.
+
+Add support for the new lint:
+```rust
+#![nit(deprecated)]
+```
+```console
+$ rustc --nit deprecated ...
+```
+```console
+$ rustc -Ndeprecated ...
+```
+
+These will be rendered like other lint levels with a diagnostic level and coloring that conveys the non-blocking, helpful nature.
+For example, this could be rendered like a `note:` or `help:` diagnostic level.
+
+Rustc will have a dynamic lint group of all `nit`s, much like `warnings` is all lints at level `warn` at the time of processing, `-Anits`.
+
+## Cargo
+
+Cargo implements the semantics for this to be a first-class non-blocking lint level.
+
+Add support for the new lint:
+```toml
+[workspace.lints.rust]
+deprecated = "nit"
+```
+
+A new config field will be added to mirror `build.warnings`:
+
+### `build.nits`
+
+- Type: string
+- Default: `allow`
+- Environment: `CARGO_BUILD_NITS`
+
+Controls how Cargo handles nits. Allowed values are:
+
+- `nit`: warnings are emitted as warnings.
+- `allow`: warnings are hidden (default).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Rust-analyzer might be limited in what it can do through LSP to avoid overwhelming the user with `nit`s.
+
+This requires a non-standard CI setup to report these messages.
+
+Non-rust-analyzer users won't get feedback until CI runs and then will have to know to run their linter in a special way to reproduce CI's output locally.
+
+Users running `CARGO_BUILD_NITS=nit cargo clippy` will have a hard time finding what lints are relevant for their change.
+If we had a way to filter lints by part of the API or by "whats new since run X",
+then that could be resolved.
+
+As `nit`s will always be enabled in the linter, just silenced in Cargo, the performance hit of running
+them will be present unless `RUSTFLAGS=-Anits` is applied by the user.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- `CARGO_BUILD_NITS=allow` is the default to avoid "warning fatigue"
+- The default level for lints is left unchanged by this RFC to keep the scope of what is designed and reviewed to the minimum
+- Rustc shows `nit`s by default, rather than hide them and require a separate opt-in mechanism for Cargo to see them
+  - Problem: People directly using rustc or using a third-party build tool may be inundated with these
+  - Lint levels may change without breaking compatibility
+  - If stabilize with no `nit`s or only downgrading `warn`s to `nit`s, then there will be no difference in the amount of output
+  - We already have a mechanism, let's not invent a new one
+- `nits` lint group is created to give rustc-only or third-party build tools a built-in way to control these besides ignoring them like Cargo
+- Cargo could print a message that `nit`s are present to help raise awareness of how to show them but they will likely always be present, so this is noise to experienced users.
```

Comment:
FWIW, in Clippy we thought about adding something similar for lints that weren't emitted because of a (default) clippy config. For example if `avoid_breaking_exported_api` is defaulted to `true`, then some lints do not get emitted. We thought about emitting a message like

```  
`avoid_breaking_exported_api = true` prevented 5 lints to be emitted
```

This would help us raising awareness for Clippy configuration options. 

But we didn't really pursue that, because OTOneH it could be noisy, as stated here, and OTOtherH I/we weren't sure if `rustc` supports emitting such things as one-liners similar to what is emitted at the end of `cargo build/check`:

```
warning: `abc` (bin "abc") generated 1 warning
```

---

Long story short: Clippy and this might both benefit from a way to emit additional metadata/information after `cargo check/clippy` to inform the user about configuration options.

Reply:
Had some long debates with someone about the right default for `avoid_breaking_exported_api` (presume established codebase / library or optimize for new user writing an application)

In a way, it could be nice to have lints silenced by `avoid_breaking_exported_api` to instead to `nit`, rather than `allow` though that feels like there might be some circumstances where that isn't a good idea.

> Long story short: Clippy and this might both benefit from a way to emit additional metadata/information after cargo check/clippy to inform the user about configuration options.

This is implementing support directly in Cargo, so it isn't quite the same as trying to get clippy to report something like this.

However, there might be a way to do this today, with Cargo's cooperation.  We were talking about cargo tracking [unused-crate-dependencies](https://doc.rust-lang.org/nightly/rustc/lints/listing/allowed-by-default.html#unused-crate-dependencies) and deciding when to report it back based on the if all crates in a package agree (if enough crates were built).   I can't remember if we were just going to key off of the lint name in json or if there was some lint-specific information that was being communicated.  If the latter, then maybe something similar can be done here and cargo can count and report it.

However, there is still the question of "is this worth it".  Also, if we do make these `nit`s instead of `allow`s, then this problem goes away.

Reply:
btw I went ahead and called out `avoid_breaking_exported_api` in 21f543aa.  While this RFC is not directly addressing it, I wanted to have a more complete picture of lint level workflows.